### PR TITLE
Add distributed support for StreamingDataset

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds distributed support for the StreamingDataset with several shuffling strategy.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18850.org.readthedocs.build/en/18850/

cc @borda

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/streaming/shuffle.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            intervals_per_process[replica_index].append(chunk_interval)
+
+        return chunks_per_process, intervals_per_process
+
+    def __call__(self, array: np.ndarray) -> List[int]:
+        return array.tolist()
+
+
+class MinShuffle(Shuffle):
+    """MinShuffle shuffle the items and ensure all the processes receive the same number of items."""
```
Comment (lantiga) [inline src/lightning/data/streaming/shuffle.py:77]:
```suggestion
    """MinShuffle shuffles the items and ensures all the processes receive the same number of items."""
```

File: src/lightning/data/streaming/shuffle.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            chunks_per_process[replica_index].append(chunk_index)
+            intervals_per_process[replica_index].append(chunk_interval)
+
+        return chunks_per_process, intervals_per_process
+
+    def __call__(self, array: np.ndarray) -> List[int]:
+        return array.tolist()
+
+
+class MinShuffle(Shuffle):
```
Comment (lantiga) [inline src/lightning/data/streaming/shuffle.py:76]: I don't think the name is explanatory - `Min` with respect to what? It makes it hard for users to decide what to do.
Also the docstring is exactly the same as `AcrossChunkShuffle`.

Honestly I would just go for no shuffle or what we call "min" shuffle today.
If I understand it well, the across chunks would be quite inefficient and it's almost an antipattern for streaming IMO.

Shuffling over one or just a few chunks is really what we should push users to choose.

File: src/lightning/data/streaming/shuffle.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+        return chunks_per_process, intervals_per_process
+
+    def __call__(self, array: np.ndarray) -> List[int]:
+        assert self.random_state
+        return self.random_state.permutation(array).tolist()
+
+
+class AcrossChunkShuffle(Shuffle):
+    """AcrossChunkShuffle shuffle the items and ensure all the processes receive the same number of items."""
```
Comment (lantiga) [inline src/lightning/data/streaming/shuffle.py:110]: This is a repetition of the previous docstring.

File: tests/tests_data/streaming/test_dataset.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+    assert len(dataset) == 816
+    dataset_iter = iter(dataset)
+    assert len(dataset_iter) == 816
+
+    dataloader = DataLoader(dataset, num_workers=2, batch_size=2)
+    assert len(dataloader) == 408
+
+
+def test_streaming_dataset_distributed_min_shuffle(tmpdir):
```
Comment (lantiga) [inline tests/tests_data/streaming/test_dataset.py:54]: this test doesn't test the effect of shuffling, only that the end result is correctly partitioned

File: tests/tests_data/streaming/test_dataset.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                break
+        found_list.append(found)
+
+    assert all(found_list) is True
+
+    assert len([i for i in process_1_1 if i in process_2_1]) == 0
+    assert len([i for i in process_1_2 if i in process_2_2]) == 0
+
+
+def test_streaming_dataset_distributed_across_chunks_shuffle(tmpdir):
```
Comment (lantiga) [inline tests/tests_data/streaming/test_dataset.py:168]: same comment as before, the test doesn't test the actual shuffling strategy

**Review by lantiga — APPROVED** [review]
Approved for expediting, but comments need to be addressed (here on in a subsequent PR)

File: src/lightning/data/streaming/dataset.py [hunk 6]
```diff
@@ -68,26 +77,16 @@ def __init__(
         self.has_triggered_download = False
         self.min_items_per_replica: Optional[int] = None
         self.seed = seed
-        self.num_iter = 0
+        self.current_epoch = 0
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:80]: This dataset isn't meant to be used by padding indexes. This is only for debugging purposes. The intuitive path is either to loop over or to use a DataLoader. 

I am not particularly worry about this right now.

File: src/lightning/data/streaming/shuffle.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:76]: We need to support both. The default is the `MinShuffle` but we can't not offer a shuffling solution that doesn't lose that many samples.

Comment (lantiga) [reply src/lightning/data/streaming/shuffle.py:76]: I get it now, yes makes sense.
Let’s maybe just find a better name for Min.
Also AcrossChunk is misleading IMO

File: src/lightning/data/streaming/shuffle.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py:77]: done !

File: src/lightning/data/streaming/shuffle.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/shuffle.py]: done.

File: tests/tests_data/streaming/test_dataset.py (hunk 5 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py]: add some extra bits.

File: tests/tests_data/streaming/test_dataset.py (hunk 4 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py:54]: Added some extra bits.

File: src/lightning/data/streaming/shuffle.py [hunk 7]
```diff
@@ -73,8 +73,18 @@ def __call__(self, array: np.ndarray) -> List[int]:
...
+class TruncatedShuffle(Shuffle):
+    """TruncatedShuffle shuffles the chunks and associates them to the ranks.
+
+    As the number of items in a chunk varies, it is possible for a rank to end up with more or less items.
+
+    To ensure the same fixed dataset length for all ranks, we compute the minimum number of items across all ranks.
+
+    For the ranks with more items than the minimum, the remaining items are dropped.
+
+    Note: This is the fatest sampling strategy but at the cost of losing items.
```
Comment (lantiga) [inline src/lightning/data/streaming/shuffle.py:85]:
```suggestion
    Note: This is the fastest sampling strategy but at the cost of losing items.
```

File: src/lightning/data/streaming/shuffle.py [hunk 8]
```diff
@@ -106,8 +116,23 @@ def __call__(self, array: np.ndarray) -> List[int]:
...
 
 
-class AcrossChunkShuffle(Shuffle):
-    """AcrossChunkShuffle shuffle the items and ensure all the processes receive the same number of items."""
+class FullShuffle(Shuffle):
+    """FullShuffle shuffles the chunks and associates them to the ranks.
+
+    As the number of items in a chunk varies, it is possible for a rank to end up with more or less items.
+
+    To ensure the same fixed dataset length for all ranks while dropping as little items as possible,
```
Comment (lantiga) [inline src/lightning/data/streaming/shuffle.py:124]:
```suggestion
    To ensure the same fixed dataset length for all ranks while dropping as few items as possible,
```

File: src/lightning/data/streaming/shuffle.py [hunk 9]
```diff
@@ -106,8 +116,23 @@ def __call__(self, array: np.ndarray) -> List[int]:
...
+
+    we adopt the following strategy.
+
+    We compute the maximum number of items per rank (M) and iterate through the chunks and ranks
+
+    until we have associated at least M items per rank.
+
+    As a result, we lose at most world_size items. However, as some chunks are shared across ranks. This leads to the
```
Comment (lantiga) [inline src/lightning/data/streaming/shuffle.py:132]:
```suggestion
    As a result, we lose at most (number of ranks) items. However, as some chunks are shared across ranks. This leads to the
```

File: src/lightning/data/streaming/shuffle.py [hunk 10]
```diff
@@ -106,8 +116,23 @@ def __call__(self, array: np.ndarray) -> List[int]:
...
+    same chunk to be downloaded multiple times
```
Comment (lantiga) [inline src/lightning/data/streaming/shuffle.py:133]:
```suggestion
    same chunk to be downloaded multiple times.
```

**Review by lantiga — APPROVED** [review]
LGTM!

File: src/lightning/data/streaming/shuffle.py (hunk 8 above)
Comment (mestir) [reply src/lightning/data/streaming/shuffle.py]: - [ ]
